stdlib/src/collections/list.mojo

```diff
@@ -197,6 +197,56 @@ struct List[T: CollectionElement](
     # Operator dunders
     # ===-------------------------------------------------------------------===#
 
+    fn __eq__[T2: ComparableCollectionElement](self: List[T], other: List[T2]) -> Bool:
+        """Checks if two lists are equal.
+
+        ```mojo
+        var x = List[Int](1, 2, 3)
+        var y = List[Int](1, 2, 3)
+        if x == y: print("x and y are equal")
+        ```
+        Parameters:
+            T2: The type of the elements in the list. Must implement the
+              traits `EqualityComparable` and `CollectionElement`.
+
+        Args:
+            other: The list to compare with.
+
+        Returns:
+            True if the lists are equal, False otherwise.
+        """
+        constrained[_type_is_eq[T, T2](), "other type is not self.T"]()
+        if len(self) != len(other):
+            return False
+        var index = 0
+        for i in self:
```

Comment:
Ideally it would be better to use `range(len(self))` here, but the iterator here returns with a `Reference` rather than a `ref` for each element which I can rebind to `Reference[T2]`. 

Perhaps there is a nicer way, but I am just not aware of that at the moment.

Reply:
does it work if you try this ?
```mojo
for i in range(len(self)):
    var left = rebind[T2](self.unsafe_get(i))
    var right = other.unsafe_get(i)
```

Reply:
@martinvuyk This wouldn't work if we can't use the one-typed formation of conditional conformance because `rebind` only works on reg-types.

Reply:
I was actually exploring changing List and Dict iterator's `__next__` to return a `ref` yesterday.  We can't quite do that yet.  In the `List` case, it actually introduces extra copies.  Note that, independent of the `ref` vs `Reference` stuff, we may be able to remove some crazy rebinds that happen in `__eq__`, `__ne__`, and `__contains__` with conditional conformance - we should explore that.  In today's nightly release, you'll see I changed `List.__contain__` to remove the rebind and just be:

```
 fn __contains__[U: ComparableCollectionElement](self: List[U], value: U) -> Bool:
   for i in self:
       if i[] == value:
           return True
    return False
```

Reply:
Thanks, going to rebase on it then.

Reply:
Updated along with using `EqualityComparableCollectionElement`, it is much nicer now, thanks!